sdk/cosmos/azure-cosmos-benchmark/src/main/java/com/azure/cosmos/benchmark/BenchmarkRequestSubscriber.java

```diff
@@ -15,26 +18,50 @@
 
 public class BenchmarkRequestSubscriber<T> extends BaseSubscriber<T> {
     final static Logger logger = LoggerFactory.getLogger(BenchmarkRequestSubscriber.class);
+    private final StopWatch durationTimer = new StopWatch();
     private Meter successMeter;
     private Meter failureMeter;
     private Semaphore concurrencyControlSemaphore;
     private AtomicLong count;
+    private int thresholdForDiagnosticsInMs;
     public Timer.Context context;
 
-    public BenchmarkRequestSubscriber(Meter successMeter, Meter failureMeter, Semaphore concurrencyControlSemaphore,  AtomicLong count) {
+    public BenchmarkRequestSubscriber(Meter successMeter,
+                                      Meter failureMeter,
+                                      Semaphore concurrencyControlSemaphore,
+                                      AtomicLong count,
+                                      int thresholdForDiagnosticsInMs) {
         this.successMeter = successMeter;
         this.failureMeter = failureMeter;
         this.concurrencyControlSemaphore = concurrencyControlSemaphore;
         this.count = count;
+        this.thresholdForDiagnosticsInMs = thresholdForDiagnosticsInMs;
     }
 
     @Override
     protected void hookOnSubscribe(Subscription subscription) {
         super.hookOnSubscribe(subscription);
+        durationTimer.start();
     }
 
     @Override
     protected void hookOnNext(T value) {
+        if (durationTimer.isStarted()) {
+            durationTimer.stop();
+        }
+
+        if (durationTimer != null && durationTimer.getTime() > thresholdForDiagnosticsInMs) {
```

Comment:
StopWatch can be used on different threads and hence it is not thread safe.
Please use a volatile long for measuring the checkpoint.


```suggestion
        if (durationTimer != null && System.currentMilis() startTimeInMs-  > thresholdForDiagnosticsInMs) {

```

Reply:
Wow - thanks - I would never have imagined StopWatch to not be thread-safe in Java

Reply:
per documentation it is not:
http://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/time/StopWatch.html
https://guava.dev/releases/18.0/api/docs/com/google/common/base/Stopwatch.html

Reply:
Done , moved to volatile long startTimeInMs , but we could have done with StopWatch with synchronous block as we are doing in gone and retry policy. Also we are creating new subscriber every time and not sharing with other request , so not sure we would have the threading problem here. Any way I am fine either way